python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -528,100 +562,275 @@ def asset_key_for_input(self, input_name: str) -> AssetKey:
-        """Returns the asset partition key for the given output. Defaults to "result", which is the
-        name of the default output.
+    def partition_key_for_asset(
+        self, asset: CoercibleToAssetKey, is_dependency: bool = False
+    ) -> Union[str, MultiPartitionKey]:
+        """Returns the partition key of the provided asset.
+
+        Args:
+            asset (Union[AssetKey, str]): The asset to get the partition key for
+            is_dependency (bool): If the asset is a self-dependent asset, you can set is_dependency=True to
+                fetch the partition key of the asset as an upstream dependency
+
+        Examples:
+            .. code-block:: python
+
+                partitions_def = DailyPartitionsDefinition("2023-08-20")
+
+                @asset(
+                    partitions_def=partitions_def
+                )
+                def upstream_asset(context: AssetExecutionContext):
+                    context.log.info(context.partition_key_for_asset("upstream_asset"))
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   "2023-08-21"
+
+                @asset(
+                    partitions_def=partitions_def,
+                )
+                def downstream_asset(context: AssetExecutionContext, upstream_asset):
+                    context.log.info(context.partition_key_for_asset("upstream_asset"))
+                    context.log.info(context.partition_key_for_asset("downstream_asset"))
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   "2023-08-21"
+                #   "2023-08-21"
+
+                @asset(
+                    partitions_def=partitions_def,
+                    ins={
+                        "self_dependent_asset": AssetIn(partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1))
+                    }
+                )
+                def self_dependent_asset(context: AssetExecutionContext, self_dependent_asset):
+                    context.log.info(context.partition_key_for_asset("self_dependent_asset", is_dependency=True))
+                    context.log.info(context.partition_key_for_asset("self_dependent_asset"))
+
+                # materializing the 2023-08-21 partition of this asset will log:
+                #   "2023-08-20"
+                #   "2023-08-21"
+
         """
-        return self._step_execution_context.asset_partition_key_for_output(output_name)
+        return self._step_execution_context.partition_key_for_asset(
+            asset, is_dependency=is_dependency
+        )
 
     @public
-    def asset_partitions_time_window_for_output(self, output_name: str = "result") -> TimeWindow:
-        """The time window for the partitions of the output asset.
+    def partitions_time_window_for_asset(
+        self, asset: CoercibleToAssetKey, is_dependency: bool = False
```

Comment:
I don't think these should take ` CoercibleToAssetKey` as an `AssetsDefinition` with multiple keys would have to be a hard error.

Reply:
I think cases like this present a pretty good argument that `CoercibleToAssetKey` shouldn't be expanded to take `AssetsDefinition` (as discussed [here](https://github.com/dagster-io/dagster/pull/16207#discussion_r1318450487)) and that a new type alias should be created to include `AssetsDefinition`